src/sentry/static/sentry/app/views/ruleEditor/utils.jsx

```diff
@@ -0,0 +1,56 @@
+import React from 'react';
+import {Select2Field} from '../../components/forms';
+
+function getChoiceField(name, data) {
+  return (
+    <Select2Field
+      name={name}
+      choices={data.choices}
+      key={name}
+      style={{marginBottom: 0}}
+    />
+  );
+}
+
+function getInputField(type, name, data) {
+  return (
+    <input
+      name={name}
+      type={type}
+      placeholder={data.placeholder}
+      key={name}
+      style={{height: 37}}
+    />
+  );
+}
+
+const getFieldTypes = {
+  choice: getChoiceField,
+  number: getInputField.bind(null, 'number'),
+  string: getInputField.bind(null, 'text'),
+};
+
+function getField(name, data) {
+  return getFieldTypes[data.type](name, data);
+}
+
+export function getComponent(node) {
+  const {label, formFields} = node;
+  const pattern = /{\w+}/g;
+  const res = [];
+  const text = label.split(pattern);
+  const matches = (label.match(pattern) || []).map(match => {
+    const key = match.slice(1, -1);
+    const fieldData = formFields[key];
+    return fieldData ? getField(key, fieldData) : match;
+  });
+
+  while (text.length) {
+    res.push(text.shift());
+    if (matches.length) {
+      res.push(matches.shift());
+    }
+  }
+
+  return res;
```

Comment:
I think you could shorten this up! How about this?

```js
const {label, formFields} = node;

return test.split(/({\w+})/).map(part => {
  if (!part.test(/^{\w+}$/)) {
    return part;
  }
  
  const key = part.slice(1, -1);
  return formFields[key]; getField(key, formFields[key]) : part
})
```

Reply:
Or alternatively using `tct` to replace the `{}` (which would have to change to `[]`) does something similar I think.